### PR TITLE
feat: Implement goto_declaration support

### PR DESCRIPTION
**Pull request by Veykril**
This is just a simple implementation that falls back to `goto_definition` for everything but modules where it goes to the actual module declaration if possible.

### CODE REVIEW

**Comment by Veykril** [comment]
For some odd reason running the tidy tests doesn't regenerate the feature docs for me currently?

**Comment by matklad** [comment]
I wonder if goto_definition fallback logically belongs to the client?

**Comment by matklad** [comment]
Another natural candidate here are trait methods.

**Comment by Veykril** [comment]
> I wonder if goto_definition fallback logically belongs to the client?

Good point, it feels like it could belong to the client.

> Another natural candidate here are trait methods.

Oh right, I was wondering if there was something else that has the notion of declaration vs definition in rust 👍

**Comment by kjeremy** [comment]
See #2541 for thoughts on this.

File: crates/ide/src/goto_declaration.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        };
+        match def {
+            Definition::ModuleDef(hir::ModuleDef::Module(module)) => Some(RangeInfo::new(
+                original_token.text_range(),
+                vec![NavigationTarget::from_module_to_decl(db, module)],
+            )),
+            _ => return None,
+        }
+    })();
+    res.or_else(|| goto_definition::goto_definition(db, position))
```
Comment (matklad) [inline crates/ide/src/goto_declaration.rs:47]: Let’s put this or_else in the handlers for now, otherwise, LGTM!

**Comment by Veykril** [comment]
bors r=matklad

**Pull request merged**